ntoskrnl/ke/i386/kiinit.c

```diff
@@ -392,6 +340,122 @@ KiInitializePcr(IN ULONG ProcessorNumber,
     Pcr->PrcbData.MultiThreadProcessorSet = Pcr->PrcbData.SetMember;
 }
 
+static
+CODE_SEG("INIT")
+VOID
+KiVerifyCpuFeatures(PKPRCB Prcb)
+{
+    CPU_INFO CpuInfo;
+
+    // 1. Check CPUID support
+    ULONG EFlags = __readeflags();
+
+    /* XOR out the ID bit and update EFlags */
+    ULONG NewEFlags = EFlags ^ EFLAGS_ID;
+    __writeeflags(NewEFlags);
+
+    /* Get them back and see if they were modified */
+    NewEFlags = __readeflags();
+
+    if (NewEFlags == EFlags)
+    {
+        /* The modification did not work, so CPUID is not supported. Try Windows 95. */
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0, 0, 0, 0); // TODO: what to put here?
+    }
+    else
+    {
+        /* CPUID is supported. Set the ID Bit again. */
+        EFlags |= EFLAGS_ID;
+        __writeeflags(EFlags);
+    }
+
+    /* Peform CPUID 0 to see if CPUID 1 is supported */
+    KiCpuId(&CpuInfo, 0);
+    if (CpuInfo.Eax == 0)
+    {
+        /* Maybe Windows 95 after all? */
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0, 0, 0, 0); // TODO: what to put here?
+    }
+
+    // 2. Detect and set the CPU Type
+    KiSetProcessorType();
+
+    // We recommend Windows 3.11 for i386 CPU.
+    if (Prcb->CpuType == 3)
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR, 0x386, 0, 0, 0);
+
+    // 3. Finally, obtain CPU features.
+    ULONG FeatureBits = KiGetFeatureBits();
+
+    // 4. Verify it supports everything we need.
+    if (!(FeatureBits & KF_RDTSC))
+    {
+        ULONG Vendor[3];
+
+        RtlCopyMemory(Vendor, Prcb->VendorString, sizeof(Vendor));
+
+        KeBugCheckEx(UNSUPPORTED_PROCESSOR,
+                     (1 << 24 ) | (Prcb->CpuType << 16) | Prcb->CpuStep,
```

Comment:
```suggestion
                     (1 << 24) | (Prcb->CpuType << 16) | Prcb->CpuStep,
```
(here and below in the 3 other copypasted lines).

Reply:
Do you think there is one of those four bugcheck fields that could be sacrified and used instead as a hint to which feature bit support is missing?

Reply:
I'd say vendor field is not that much important at all :) The best way is to indicate which feature we didn't like here, yes